examples/advanced/llm_hf/README.md

```diff
@@ -0,0 +1,137 @@
+
+## Setup
+Please make sure you set up virtual environment following [example root readme](../../README.md).
+Install additional requirements:
+```
+python3 -m pip install -r requirements.txt
+```
+
+## Model and Data Preparation
+We first download the model and save it to the `model` folder, note that approved access to the model is required
+```
+mkdir model
+cd model
+git clone https://huggingface.co/meta-llama/Llama-2-7b-hf
+cd ..
+```
+
+We then download and preprocess (to be consistent with our [NeMo example](../../../integration/nemo/examples/supervised_fine_tuning), we follow the same preprocessing steps) the dataset [databricks-dolly-15k](https://huggingface.co/datasets/databricks/databricks-dolly-15k) for this example. 
+```
+mkdir dataset
+cd dataset
+git clone https://huggingface.co/datasets/databricks/databricks-dolly-15k
+cd ..
+mkdir dataset/dolly
+python ./utils/preprocess_dolly.py --training_file dataset/databricks-dolly-15k/databricks-dolly-15k.jsonl --output_dir dataset/dolly
+```
+
+## Centralized Training
+### One-call training
+Centralized trainings, as the baseline for comparison with FL results, are done with the following command:
+```
+python3 ./utils/hf_sft_peft.py --output_path ./workspace_centralized/llama2-7b-dolly-sft --mode 0
+python3 ./utils/hf_sft_peft.py --output_path ./workspace_centralized/llama2-7b-dolly-peft --mode 1
+```
+### Adaptation Step 1: iterative training
+To adapt the centralized training script to federated application, under ``launch_once'' setting, we first need to "break" the single call to `trainer.train()` into iterative calls, one for each round of training.
+For this purpose, we provided `utils/hf_sft_peft_iter.py` as an example, which is a modified version of `utils/hf_sft_peft.py`.
+Their differences are highlighted below:
+
+![diff](./figs/diff_1.png)
+![diff](./figs/diff_2.png)
+
+Note that the `trainer.train()` call is replaced by a `for` loop. At the beginning of each round (epoch in this case), we intentionally load a fixed model weights saved at the beginning, over-write the last round's record, then call `trainer.train(resume_from_checkpoint=True)` with `trainer.args.num_train_epochs` incremented by 1. 
+
+The purpose of doing so is to tell if the intended weights are succesfully loaded at each round. Without using a fixed starting model, even if the model weights are not properly loaded, the training loss curve will still follow the one-call result, which is not what we want to see. 
+
+If the intended model weights (serving as the starting point for each round, the "global model" for FL use case) is properly loaded, then we shall observe a "zig-zag" pattern in the training loss curve because the model weights are reset to the same starting point at the beginning of each round. In contrast to the one-shot centralized training, where the model weights are updated continuously, and the training loss curve should follow an overall decreasing trend.
```

Comment:
Insert "This is because..." before "In contrast..."

Reply:
added